dndsim/src/classes/Barbarian.ts

```diff
@@ -0,0 +1,254 @@
+import { ClassLevel } from "../sim/coreFeats/ClassLevel"
+import { applyFeatSchedule, defaultMagicBonus } from "../util/helpers"
+import { ExtraAttack } from "../feats/shared/ExtraAttack"
+import { Resource } from "../sim/resources/Resource"
+import { Greatsword } from "../weapons"
+import { SavageAttacker } from "../feats/origin/SavageAttacker"
+import { GreatWeaponMaster } from "../feats/general/GreatWeaponMaster"
+import { AbilityScoreImprovement } from "../feats/general/AbilityScoreImprovement"
+import { IrresistibleOffense } from "../feats/epic/IrresistibleOffense"
+import { SetAttribute } from "../feats/shared/SetAttribute"
+import { IncreaseResource } from "../feats/shared/IncreaseResource"
+import { WeaponAttack } from "../sim/Attack"
+
+const RageResource = "rage"
+const RageEffect = "raging"
+const RecklessTag = "reckless"
+
+class RageOperation implements Operation {
+    repeatable = false
+
+    eligible(_environment: Environment, character: Character): boolean {
+        // While it is techically possible to rage while raging, at this time
+        // there is no benefit to doing so
+        return character.hasResource(RageResource)
+            && character.bonus.has()
+            && !character.hasEffect(RageEffect)
+    }
+
+    do(_environment: Environment, character: Character): void {
+        character.useResource(RageResource)
+        character.bonus.use()
+        character.addEffect(RageEffect)
+    }
+}
+
+const RageBonusDamageAttribute = "rageBonusDamage"
+
+class Rage extends Feat {
+    apply(character: Character) {
+        this.addResource()
+        character.customTurn.addOperation("before_action", new RageOperation())
+        character.events.on("damage_roll", (event) => this.damageRoll(event))
+        character.events.on("short_rest", () => this.shortRest())
+    }
+
+    damageRoll(event: DamageRollEvent) {
+        if (event.attack?.attack?.weapon()?.mod(this.character) == "str") {
+            event.damage.flatDmg += this.character.getAttribute(RageBonusDamageAttribute);
+        }
+    }
+
+    shortRest() {
+        this.character.removeEffect(RageEffect)
+    }
+
+    addResource() {
+        this.character.resources.set(
+            RageResource,
+            new Resource({
+                name: RageResource,
+                character: this.character,
+                initialMax: 2,
+                incrementOnShortRest: true,
+                resetOnLongRest: true,
+
+            })
+        )
+    }
+}
+
+class AddRage extends Feat {
+    apply(character: Character): void {
+        character.getResource(RageResource).addMax(1)
+    }
+}
+
+class RecklessAttack extends Feat {
+    apply(character: Character) {
+        character.events.on("attack_roll", (event) => this.attackRoll(character, event))
+    }
+
+    attackRoll(character: Character, event: AttackRollEvent) {
+        if (event.attack?.attack?.weapon()?.mod(character) == "str") {
+            event.adv = true
```

Comment:
Implementing Brutal Strikes correctly may require making `adv` more than just a boolean. With the current system, I need to either set both `adv` and `disadv` to `true`, or both to `false`, to make it a straight roll. If I set both to `true`, then we can't capture that if there are any real sources of disadvantage then the Brutal Strike can't apply. If I set both to `false`, then you could get additional sources of advantage or disadvantage, and they wouldn't be canceled out.

Reply:
Naming it `after_attack_roll` would be confusing, since you're still applying something that occurs before the roll happens.

I think it would be cleanest to implement if we made advantage and disadvantage lists of reasons instead of just booleans. We would give it the `RecklessAttack` reason for advantage, and the `BrutalStrike` reason for disadvantage. Then, at the damage step (or any step where we applied something from Brutal Strike) we'd check the list of disadvantage reasons. If it's only `BrutalStrike`, we apply. If it has anything else, we don't apply.